pynapple/process/signal_processing.py

```diff
@@ -0,0 +1,267 @@
+"""
+Signal processing tools for Pynapple.
+
+Contains functionality for signal processing pynapple object; fourier transforms and wavelet decomposition.
+"""
+
+from itertools import repeat
+from math import ceil, floor
+
+import numpy as np
+import pandas as pd
+from scipy.signal import welch
+
+import pynapple as nap
+
+
+def compute_spectogram(sig, fs=None):
+    """
+    Performs numpy fft on sig, returns output
+
+    ----------
+    sig : pynapple.Tsd or pynapple.TsdFrame
+        Time series.
+    fs : float, optional
+        Sampling rate, in Hz. If None, will be calculated from the given signal
+    """
+    if not isinstance(sig, (nap.Tsd, nap.TsdFrame)):
+        raise TypeError(
+            "Currently compute_spectogram is only implemented for Tsd or TsdFrame"
+        )
+    if fs is None:
+        fs = sig.index.shape[0] / (sig.index.max() - sig.index.min())
+    fft_result = np.fft.fft(sig.values, axis=0)
+    fft_freq = np.fft.fftfreq(len(sig.values), 1 / fs)
+    return pd.DataFrame(fft_result, fft_freq)
+
+
+def compute_welch_spectogram(sig, fs=None):
+    """
+    Performs scipy Welch's decomposition on sig, returns output
+
+    ----------
+    sig : pynapple.Tsd or pynapple.TsdFrame
+        Time series.
+    fs : float, optional
+        Sampling rate, in Hz. If None, will be calculated from the given signal
+    """
+    if not isinstance(sig, (nap.Tsd, nap.TsdFrame)):
+        raise TypeError(
+            "Currently compute_welch_spectogram is only implemented for Tsd or TsdFrame"
+        )
+    if fs is None:
+        fs = sig.index.shape[0] / (sig.index.max() - sig.index.min())
+    freqs, spectogram = welch(sig.values, fs=fs, axis=0)
+    return pd.DataFrame(spectogram, freqs)
+
+
+def _morlet(M=1024, ncycles=1.5, scaling=1.0, precision=8):
+    """
+    Defines the complex Morelet wavelet kernel
+
+    Parameters
+    ----------
+    M : int
+        Length of the wavelet
+    ncycles : float
+        number of wavelet cycles to use. Default is 5
+    scaling: float
+        Scaling factor. Default is 1.5
+    precision: int
+        Precision of wavelet to use
+
+    Returns
+    -------
+    np.ndarray
+        Morelet wavelet kernel
+    """
+    x = np.linspace(-precision, precision, M)
+    return (
+        ((np.pi * ncycles) ** (-0.25))
+        * np.exp(-(x**2) / ncycles)
+        * np.exp(1j * 2 * np.pi * scaling * x)
+    )
```

Comment:
If you store the morlet kernels in a (2, k) array (first axis is real/imagm, second axis is the number of kernels you have). you can call `nap.convolve` over it later on. `pynapple` will handle the convolution over both column dealing with the disjoint epochs.

this will have the additional benefit to work out of the box with pynajax